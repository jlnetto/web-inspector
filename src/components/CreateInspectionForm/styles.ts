import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
`;

export const Title = styled.h2`
  font-size: 20px;
  margin: 0;
`;

export const FormWrapper = styled.form`
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  gap: 15px;
`;
