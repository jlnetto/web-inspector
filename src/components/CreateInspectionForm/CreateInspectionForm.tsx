import { useState, useEffect } from 'react';
import axios from 'axios';
import useSWR from 'swr';
import { createInspection } from '../../service/api/endpoints';
import type { StoredData } from '../../types/core';
import updateLocalStorage from '../../utils/updateLocalStorage';

import Button from '../Button';
import TextField from '../TextField';
import Banner from '../Banner';

import { Container, Title, FormWrapper } from './styles';

type BannerConfig = {
  type: 'error' | 'success';
  message: string;
};

const CreateInspectionForm = () => {
  const [keyword, setKeyword] = useState<string>('');
  const [bannerConfig, setBannerConfig] = useState<BannerConfig>({
    type: 'success',
    message: 'Success!',
  });
  const [showBanner, setShowBanner] = useState<boolean>(false);
  const [lastKeywords, setLastKeywords] = useState<StoredData[]>([]);

  const { mutate, isLoading } = useSWR('/crawl');

  useEffect(() => {
    const storedKeywords = window.localStorage.getItem('lastKeywords');
    if (storedKeywords) {
      setLastKeywords(JSON.parse(storedKeywords));
    }
  }, []);

  useEffect(() => {
    if (!lastKeywords.length) return;
    updateLocalStorage('lastKeywords', JSON.stringify(lastKeywords));
  }, [lastKeywords]);

  const handleMessage = ({
    type,
    message,
  }: {
    type: 'error' | 'success';
    message: string;
  }) => {
    setShowBanner(true);
    setBannerConfig({
      type,
      message,
    });
  };

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();

    try {
      const { data } = await createInspection(keyword);
      setKeyword('');
      setLastKeywords((prevKeywords) => [
        ...prevKeywords,
        { id: data.id, keyword },
      ]);
      mutate();
      handleMessage({
        type: 'success',
        message: 'Cadastro de uma solicitação de inspeção realizada com sucesso!',
      });
    } catch (error) {
      if (axios.isAxiosError(error)) {
        handleMessage({
          type: 'error',
          message: `Internal error: ${error.response?.data.message}`,
        });
      }
      console.error(error);
    }
  };

  return (
    <Container>
      <Title>Cadastro de uma solicitação de inspeção</Title>
      {showBanner ? (
        <Banner
          type={bannerConfig.type}
          message={bannerConfig.message}
          onClose={() => setShowBanner(false)}
        />
      ) : null}
      <FormWrapper onSubmit={handleSubmit}>
        <TextField
          label="Solicitação de inspeção"
          placeholder='Palavra-chave'
          name="register"
          value={keyword}
          onChange={setKeyword}
        />
        <Button
          type="submit"
          text="Enviar"
          disabled={keyword.length === 0}
          loading={isLoading}
        />
      </FormWrapper>
    </Container>
  );
};

export default CreateInspectionForm;
