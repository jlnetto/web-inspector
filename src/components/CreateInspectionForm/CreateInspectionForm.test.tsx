import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import CreateInspectionForm from './CreateInspectionForm';

describe('CreateInspectionForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render form correctly', () => {
    const { asFragment } = render(<CreateInspectionForm />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should submit the form and display success message', async () => {
    render(<CreateInspectionForm />);

    const keywordInput = screen.getByRole('textbox');
    fireEvent.change(keywordInput, { target: { value: 'Keyword' } });

    const submitButton = screen.getByText('Enviar');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(
        screen.getByText(
          'Cadastro de uma solicitação de inspeção realizada com sucesso!'
        )
      ).toBeInTheDocument();
    });
  });
});
