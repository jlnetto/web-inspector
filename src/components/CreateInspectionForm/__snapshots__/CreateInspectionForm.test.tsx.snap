// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateInspectionForm should render form correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-ispOId gJMkEM"
  >
    <h2
      class="sc-kcuKUB ckVtbx"
    >
      Cadastro de uma solicitação de inspeção
    </h2>
    <form
      class="sc-dcrmVg cZdKfx"
    >
      <div
        class="sc-fFGjHI boHwgv"
      >
        <label
          for="register"
        >
          Solicitação de inspeção
        </label>
        <input
          class="sc-ewnqHT hGfAjc"
          placeholder="Palavra-chave"
          value=""
        />
      </div>
      <button
        class="sc-bgqQcB kYYpOZ"
        disabled=""
        type="submit"
      >
        Enviar
      </button>
    </form>
  </div>
</DocumentFragment>
`;
